File: October 22/08 Oct/Substrings of length k with k-1 distinct elements gfg.cpp
    int countOfSubstrings(string S, int K) {
        // code here
        int n=S.length();
        vector<int> window(26,0);
        for(int i=0;i<K;i++)
        {
            window[S[i]-'a']++;
        }
        int distinct=0,subString=0;
        for(int i=0;i<26;i++)
                distinct+=(window[i]>0);
                
        if(distinct==(K-1)) subString++;
        for(int i=K;i<n;i++)
        {
            window[S[i-K]-'a']--;
            window[S[i]-'a']++;
            distinct=0;
            for(int i=0;i<26;i++)
                distinct+=(window[i]>0);
            if(distinct==(K-1)) subString++;
        }
        
        return subString;
    }

File: JULY 2025/01 July/Substrings of length k with k-1 distinct elements gfg.java
class Solution {
    public int substrCount(String s, int k) {
        // code here
        int N = s.length();
        int[] windowFreq = new int[26];
        int countOfDuplicate = 0, ch;
        for(int i=0; i<Math.min(N, k); i++)
        {
            ch = s.charAt(i)-'a';
            windowFreq[ch]++;
        }
        
        for(int i=0; i<26; i++)
        {
            if(windowFreq[i]>1)
                countOfDuplicate+=windowFreq[i];
        }
        int countOfSubstr = (countOfDuplicate == 2? 1: 0);
        for(int i=k; i<N; i++)
        {
            ch = s.charAt(i)-'a';
            // if(windowFreq[ch]==1)
            //     countOfDuplicate++;
            windowFreq[ch]++; // move right end of window 
            
            ch = s.charAt(i-k)-'a';
            // if(windowFreq[ch]==2)
            //     countOfDuplicate--;
            windowFreq[ch]--; // move left end of window
            
            countOfDuplicate = 0;
            for(int j=0; j<26; j++)
            {
                if(windowFreq[j]>1)
                    countOfDuplicate += windowFreq[j];
            }

            // System.out.println("Index "+i+" dupCount = "+countOfDuplicate+" strCount = "+countOfSubstr);
            countOfSubstr += (countOfDuplicate == 2? 1: 0);
        }
        
        return countOfSubstr;
    }
}
